independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/CurrentInjectionPointProvider.java

```diff
@@ -177,14 +177,20 @@ public int getPosition() {
 
     }
 
+    private static <V> Set<V> trimImmutableSet(final Set<V> input) {
+        if (input == null)
+            return null;
+        return input.isEmpty() ? Collections.EMPTY_SET : input;
+    }
+
     static abstract class AnnotatedBase implements Annotated {
 
         private final Type baseType;
         private final Set<Annotation> annotations;
 
         AnnotatedBase(Type baseType, Set<Annotation> annotations) {
             this.baseType = baseType;
-            this.annotations = annotations;
+            this.annotations = trimImmutableSet(annotations); //Don't retain memory for empty sets: HashSet is relatively large.
```

Comment:
Stupid question: is it a case where we often have 1 element? I know for HV, it was a very common case and we decided to optimize it properly: https://github.com/hibernate/hibernate-validator/blob/master/engine/src/main/java/org/hibernate/validator/internal/util/CollectionHelper.java#L106 .

It might not be the case here so put it in the crazy idea box if it doesn't make sense :).

Reply:
that's a great idea :)

I just queried my heap dump - found many such cases within ORM, and some in Vert.x. I have no doubt it would help in Arc as well, I just don't have mny annotations in my benchmark...

```
SELECT OBJECTS inbounds(l) FROM "java.util.HashSet" l WHERE (l.map.size = 1)
```

Reply:
damn, now I have another round of patching to do in Hibernate ORM...

Reply:
actually in ORM the extra wrapping is detrimental. I'd need a "maybeimmutable" version...

Reply:
Yeah, sorry I should have been more clear: I was specifically talking about the `singleton` part. I agree that wrapping in an immutable collection might be detrimental.

Reply:
thanks - applied the suggestion from Validator and a couple more improvements.

There's likely several more similar optimisations to do here, but I need to move on.